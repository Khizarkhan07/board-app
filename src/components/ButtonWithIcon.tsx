import React from "react";


type buttonProps = {
  icon?: string ;
  text?: string ;
  color?: string ;
  callback?: (e: React.MouseEvent<HTMLButtonElement>) => void;
};

export const ButtonWIthIcon: React.FC<buttonProps> = ({
  icon,
  text,
  color,
  callback,
}) => {

  const styles = {
    background: color,
    width: '100%'
  }

  return (
    <button
      style={styles}
      onClick={callback}
      className="btn btn-light btn-outline mb-1"
    >
      {icon && <i className={icon + " mr-1"}/>}
      {text && <span>{text}</span>}
    </button>
  );
};
