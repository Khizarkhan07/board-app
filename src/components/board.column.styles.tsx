import styled from "styled-components";

type BoardColumnContentStylesProps = {
  isDraggingOver: boolean;
};

export const BoardColumnWrapper = styled.div`
  flex: 1;
  padding: 8px;
  background-color: #ebecf0;
  border-radius: 4px;

  & + & {
    margin-left: 12px;
  }
`;

export const BoardColumnTitle = styled.h6`
  color: #314261;
  margin-bottom: 12px;
`;

export const BoardColumnContent = styled.div<BoardColumnContentStylesProps>`
  min-height: 20px;
  background-color: ${(props) => (props.isDraggingOver ? "#aecde0" : null)};
  border-radius: 4px;
`;

export const AddButtonWrapper = styled.div`
  min-height: 20px;
  border-radius: 4px;
  margin-top: 5px;
  
`;
