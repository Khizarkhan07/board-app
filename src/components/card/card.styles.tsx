import styled from "styled-components";
type BoardItemStylesProps = {
  isDragging: boolean;
};
export const BoardItemEl = styled.div<BoardItemStylesProps>`
  padding: 7px;
  background-color: ${(props) => (props.isDragging ? "#d3e4ee" : "#fff")};
  border-radius: 4px;
  transition: background-color 0.25s ease-out;
  width: 100%;
  position: relative;
  display: inline-block;
  span {
    display: none;
    position: absolute;
    top: 0;
    right: 0;
  }
  div {
    font-size: 11px;
  }
  
  &:hover {
    background-color: #f7fafc;
    span {
      display: flex;
    }
    
  }

  & + & {
    margin-top: 4px;
  }
`;

export const TextAreaWrapper = styled.div`
  i {
    margin: 3px;
    float:  right;
  }
`;