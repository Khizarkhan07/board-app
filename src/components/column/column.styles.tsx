import styled from "styled-components";

type BoardColumnContentStylesProps = {
  isDraggingOver: boolean;
};

export const BoardColumnWrapper = styled.div`
  flex: 1;
  padding: 8px;
  background-color: #ebecf0;
  border-radius: 4px;

  & + & {
    margin-left: 12px;
  }
`;
export const TitleWrapper = styled.div`
  span {
    display: none;
    margin-top: -40px;
    float: right;
  }
  &:hover {
    span {
      display: flex;
    }
  }
`;
export const BoardColumnTitle = styled.h6`
  display: flex;
  color: #314261;
  margin-bottom: 15px;
 
`;

export const BoardColumnContent = styled.div<BoardColumnContentStylesProps>`
  min-height: 20px;
  background-color: ${(props) => (props.isDraggingOver ? "#aecde0" : null)};
  border-radius: 4px;
`;

export const AddButtonWrapper = styled.div`
  min-height: 20px;
  border-radius: 4px;
  margin-top: 5px;
`;
