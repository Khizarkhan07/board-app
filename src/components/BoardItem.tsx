import React, {  useState } from "react";
import {Board} from "../contexts/BoardContext";
import {
  Draggable,
  DraggableProvided,
  DraggableStateSnapshot,
} from "react-beautiful-dnd";

import { BoardItemEl, TextAreaWrapper } from "./boardItem.styles";
import { ItemObjectType } from "../types"
import { ButtonWIthIcon } from "./ButtonWithIcon";

type BoardItemProps = {
  index: number;
  item: ItemObjectType;
};

export const BoardItem: React.FC<BoardItemProps> = ({ index, item }) => {
  const { dispatch } = Board();
  const [editState, setEditState] = useState(false);
  const [itemContent, setItemContent] = useState(item.content);
  const handleEditState = (e: React.MouseEvent<HTMLButtonElement>) => {
    setEditState(true);
  };

  const handleEdit = (e: React.MouseEvent<HTMLButtonElement>) => {
    dispatch({type: 'EDIT_ITEM', payload: {item: {id: item.id, content: itemContent}}})
    setEditState(false);
  };
  return (
    <React.Fragment>
      {!editState ? (
        <Draggable draggableId={item.id} index={index}>
          {(provided: DraggableProvided, snapshot: DraggableStateSnapshot) => (
            <BoardItemEl
              {...provided.draggableProps}
              {...provided.dragHandleProps}
              ref={provided.innerRef}
              isDragging={snapshot.isDragging}
            >
              {item.content}
              <span>
                <ButtonWIthIcon
                  icon={"fa fa-pencil"}
                  callback={handleEditState}
                />
              </span>
            </BoardItemEl>
          )}
        </Draggable>
      ) : (
        <TextAreaWrapper>
          <i
            onClick={(e) => {
              setEditState(false);
            }}
            className={"fa fa-times"}
          />
          <textarea
            className="form-control mb-1 mt-1"
            value={itemContent}
            onChange={(e) => {
              setItemContent(e.target.value);
            }}
          />
          <ButtonWIthIcon
            text={"Save"}
            color={"#5aac44"}
            callback={handleEdit}
          />
        </TextAreaWrapper>
      )}
    </React.Fragment>
  );
};
