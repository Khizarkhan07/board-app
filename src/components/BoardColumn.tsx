import React, { useState } from "react";
import { Board } from "../contexts/BoardContext";
import { Droppable } from "react-beautiful-dnd";
import { nanoid } from "nanoid";
import {
  BoardColumnWrapper,
  BoardColumnContent,
  AddButtonWrapper,
  BoardColumnTitle,
} from "./board.column.styles";

import { BoardItem } from "./BoardItem";
import { ButtonWIthIcon } from "./ButtonWithIcon";
import { ColumnObjectType, ItemObjectType} from "../types";
import {TextAreaWrapper} from "./boardItem.styles";

type BoardColumnProps = {
  key: string;
  column: ColumnObjectType;
  items: ItemObjectType[];
};

export const BoardColumn: React.FC<BoardColumnProps> = (props) => {
  const { dispatch } = Board();
  const [addField, setAddField] = useState<boolean>(false);
  const [addCardField, setAddCardField] = useState<boolean>(false);
  const [title, setTitle] = useState<string>('');
  const [content, setContent] = useState<string>("");

  const addClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    setAddField(true);
  };

  const addCardClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    setAddCardField(true);
  };

  const addItem = (e: React.MouseEvent<HTMLButtonElement>) => {
    dispatch({
      type: "ADD_ITEM",
      payload: { column: props.column.id, item: { content, id: nanoid() } },
    });
    setAddField(false);
  };

  const addCard = (e: React.MouseEvent<HTMLButtonElement>) => {
    dispatch({
      type: "ADD_CARD",
      payload: { column: { title: title, id: nanoid(), itemsIds: [] } },
    });
    setAddCardField(false);
  };

  return (
    <BoardColumnWrapper>
      {props.column.title ? (
        <>
          <BoardColumnTitle>{props.column.title}</BoardColumnTitle>

          <Droppable droppableId={props.column.id}>
            {(provided, snapshot) => (
              <BoardColumnContent
                {...provided.droppableProps}
                ref={provided.innerRef}
                isDraggingOver={snapshot.isDraggingOver}
              >
                {props.items.map((item: any, index: number) => (
                  <BoardItem key={item.id} item={item} index={index} />
                ))}
                {provided.placeholder}
              </BoardColumnContent>
            )}
          </Droppable>

          {!addField ? (
            <AddButtonWrapper>
              <ButtonWIthIcon
                callback={addClick}
                icon={"fa fa-plus"}
                text={"Add another card"}

              />
            </AddButtonWrapper>
          ) : (
            <TextAreaWrapper>
              <i
                  onClick={(e) => {
                    setAddField(false);
                  }}
                  className={"fa fa-times"}
              />
              <textarea
                className="form-control mb-1"
                placeholder={"Enter a title for this card"}
                onChange={(e) => {
                  setContent(e.target.value);
                }}
              />
              <ButtonWIthIcon
                callback={addItem}
                color={"#5aac44"}
                text={"Add"}
              />
            </TextAreaWrapper>
          )}
        </>
      ) : (
        <>
          {!addCardField ? (
              <AddButtonWrapper>
                <ButtonWIthIcon
                    callback={addCardClick}
                    icon={"fa fa-plus"}
                    text={"Add card"}
                />
              </AddButtonWrapper>
          ) : (
              <TextAreaWrapper>
                <i
                    onClick={(e) => {
                      setAddCardField(false);
                    }}
                    className={"fa fa-times"}
                />
              <textarea
                  className="form-control mb-1"
                  placeholder={"Enter a title for this card"}
                  onChange={(e) => {
                    setTitle(e.target.value);
                  }}
              />
                <ButtonWIthIcon
                    callback={addCard}
                    color={"#5aac44"}
                    text={"Add"}
                />
              </TextAreaWrapper>
          )}
        </>
      )}
    </BoardColumnWrapper>
  );
};
