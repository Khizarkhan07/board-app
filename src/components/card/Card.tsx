import React, {useCallback, useState} from "react";
import {useBoard} from "../../contexts/BoardContext";
import {
  Draggable,
  DraggableProvided,
  DraggableStateSnapshot,
} from "react-beautiful-dnd";

import { BoardItemEl, TextAreaWrapper } from "./card.styles";
import {BoardColumnTitle} from "../column/column.styles";
import { ItemObjectType } from "../../types"
import { ButtonWIthIcon } from "../ButtonWithIcon";

type BoardItemProps = {
  index: number;
  item: ItemObjectType;
};

export const Card: React.FC<BoardItemProps> = ({ index, item }) => {

  const { dispatch } = useBoard();
  const [editState, setEditState] = useState(false);
  const [itemContent, setItemContent] = useState(item.content);
  const [description, setDescription] = useState(item.description);

  const handleEditState = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    setEditState(true);
  },[editState]);

  const handleDelete = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    dispatch({type: 'DELETE_ITEM', payload: {item}})
  },[item]);

  const handleEdit =  (e: React.MouseEvent<HTMLButtonElement>) => {
    if(Boolean(itemContent) && Boolean(description)){
      dispatch({type: 'EDIT_ITEM', payload: {item: {id: item.id, content: itemContent, description, updated: new Date(Date.now())}}})
      setEditState(false);
    }

  };

  const handleEditToogle = useCallback((e : React.MouseEvent<HTMLButtonElement>) => {
    setEditState(!editState)
  }, [editState])

  const handleItemContent = useCallback((e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setItemContent(e.target.value)
  }, [itemContent])

  const handleDescription = useCallback((e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setDescription(e.target.value)
  }, [description])


  return (
    <React.Fragment>
      {!editState ? (
        <Draggable draggableId={item.id} index={index}>
          {(provided: DraggableProvided, snapshot: DraggableStateSnapshot) => (
            <BoardItemEl
              {...provided.draggableProps}
              {...provided.dragHandleProps}
              ref={provided.innerRef}
              isDragging={snapshot.isDragging}
            >
              <BoardColumnTitle> {item.content} </BoardColumnTitle>
              <hr/>
              {item.description}
              <br/>
              <div className={"text-muted"}>
              {item.updated.toLocaleTimeString()}
              </div>
              <span>
                <ButtonWIthIcon
                  icon="fa fa-pencil"
                  callback={handleEditState}
                />
                <ButtonWIthIcon
                    icon="fa fa-trash"
                    callback={handleDelete}
                />
              </span>
            </BoardItemEl>
          )}
        </Draggable>
      ) : (
        <TextAreaWrapper>
          <i
            onClick={handleEditToogle}
            className="fa fa-times"
          />
          <textarea
            rows={1}
            className="form-control mb-1 mt-1"
            value={itemContent}
            onChange={handleItemContent}
          />
          <textarea
              className="form-control mb-1 mt-1"
              value={description}
              onChange={handleDescription}
          />

          <ButtonWIthIcon
            text="Save"
            color="#5aac44"
            callback={handleEdit}
          />

        </TextAreaWrapper>
      )}
    </React.Fragment>
  );
};
