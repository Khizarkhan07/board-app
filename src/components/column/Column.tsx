import React, {useCallback, useState} from "react";
import { useBoard } from "../../contexts/BoardContext";
import {Droppable, Draggable, DraggableProvided, DraggableStateSnapshot} from "react-beautiful-dnd";
import { nanoid } from "nanoid";
import {
  BoardColumnWrapper,
  BoardColumnContent,
  AddButtonWrapper,
  BoardColumnTitle,
  TitleWrapper,
} from "./column.styles";

import { Card } from "../card/Card";
import { ButtonWIthIcon } from "../ButtonWithIcon";
import { ColumnObjectType, ItemObjectType } from "../../types";
import { TextAreaWrapper } from "../card/card.styles";

type BoardColumnProps = {
  key: string;
  column: ColumnObjectType;
  items: ItemObjectType[];
  index: number;
};

export const Column: React.FC<BoardColumnProps> = (props) => {
  const { dispatch } = useBoard();
  const [addField, setAddField] = useState<boolean>(false);
  const [titleToggle, setTitleToggle] = useState<boolean>(false);
  const [addCardField, setAddCardField] = useState<boolean>(false);
  const [title, setTitle] = useState<string>("");
  const [columnTitle, setColumnTitle] = useState<string>(props.column.title);
  const [content, setContent] = useState<string>("");
  const [description, setDescription] = useState<string>("");

  const addClick = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    setAddField(true);
  },[addField]);

  const addCardClick = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    setAddCardField(true);
  }, [addCardField]);

  const handleTitleToggle = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    setTitleToggle((prevState => (!prevState)))
    setColumnTitle(props.column.title);

  }, [titleToggle]);


  const handleEditTitle = useCallback((e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setColumnTitle(e.target.value);
  }, [columnTitle])

  const handleAddCardToggle = useCallback((e: React.MouseEvent<HTMLButtonElement>) => {
    setAddField(prevState => !prevState)
  }, [addField])

  const handleChangeContent = useCallback((e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setContent(e.target.value);
  }, [content])

  const handleChangeDescription = useCallback((e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setDescription(e.target.value);
  }, [description])

  const handleChangeColTitle = useCallback((e: React.ChangeEvent<HTMLTextAreaElement>) => {
    setTitle(e.target.value);
  }, [title])
  const addItem = (e: React.MouseEvent<HTMLButtonElement>) => {
    if(Boolean(description) && Boolean(content)) {
      dispatch({
        type: "ADD_ITEM",
        payload: { column: props.column.id, item: { content, id: nanoid(), description, updated: new Date(Date.now()) } },
      });
      setAddField(false);
    }

  };
  const editColumnTitle = (e: React.MouseEvent<HTMLButtonElement>) => {
    if(Boolean(columnTitle)){
      dispatch({
        type: "EDIT_COLUMN_TITLE",
        payload: { column: props.column.id, title: { columnTitle } },
      });
      setTitleToggle(false);
    }

  };

  const addCard = (e: React.MouseEvent<HTMLButtonElement>) => {
    if(Boolean(title)){
      dispatch({
        type: "ADD_CARD",
        payload: { column: { title: title, id: nanoid(), itemsIds: [] } },
      });
      setAddCardField(false);
    }

  };


  return (

    <Draggable draggableId={props.column.id} index={props.index}>
      {(provided: DraggableProvided, snapshot: DraggableStateSnapshot) => (

          <BoardColumnWrapper
              {...provided.draggableProps}
              ref={provided.innerRef}
          >
            {props.column.title ? (
                <React.Fragment>
                  {!titleToggle ? (
                      <TitleWrapper>
                        <BoardColumnTitle
                            {...provided.dragHandleProps}>{props.column.title}</BoardColumnTitle>
                        <span>
                <ButtonWIthIcon
                    icon="fa fa-pencil"
                    callback={handleTitleToggle}
                />
              </span>
                      </TitleWrapper>
                  ) : (
                      <TextAreaWrapper>
                        <i
                            onClick={handleTitleToggle}
                            className="fa fa-times"
                        />
                        <textarea
                            className="form-control mb-1"
                            placeholder="Enter a title for this column"
                            value={columnTitle}
                            onChange={handleEditTitle}
                        />
                        <ButtonWIthIcon
                            callback={editColumnTitle}
                            color="#5aac44"
                            text="Add"
                        />
                      </TextAreaWrapper>
                  )}

                  <Droppable droppableId={props.column.id} type={"task"}>
                    {(provided, snapshot) => (
                        <BoardColumnContent
                            {...provided.droppableProps}
                            ref={provided.innerRef}
                            isDraggingOver={snapshot.isDraggingOver}
                        >
                          {props.items.map((item: any, index: number) => (
                              <Card key={item.id} item={item} index={index} />
                          ))}
                          {provided.placeholder}
                        </BoardColumnContent>
                    )}
                  </Droppable>

                  {!addField ? (
                      <AddButtonWrapper>
                        <ButtonWIthIcon
                            callback={addClick}
                            icon="fa fa-plus"
                            text="Add another card"
                        />
                      </AddButtonWrapper>
                  ) : (
                      <TextAreaWrapper>
                        <i
                            onClick={handleAddCardToggle}
                            className="fa fa-times"
                        />
                        <textarea
                            rows={1}
                            className="form-control mb-1"
                            placeholder="title"
                            onChange={handleChangeContent}
                        />
                        <textarea
                            className="form-control mb-1"
                            placeholder="description"
                            onChange={handleChangeDescription}
                        />
                        <ButtonWIthIcon callback={addItem} color="#5aac44" text="Add" />
                      </TextAreaWrapper>
                  )}
                </React.Fragment>
            ) : (
                <React.Fragment>
                  {!addCardField ? (
                      <AddButtonWrapper>
                        <ButtonWIthIcon
                            callback={addCardClick}
                            icon="fa fa-plus"
                            text="Add Column"
                        />
                      </AddButtonWrapper>
                  ) : (
                      <TextAreaWrapper>
                        <i
                            onClick={(e) => {
                              setAddCardField(false);
                            }}
                            className="fa fa-times"
                        />
                        <textarea
                            className="form-control mb-1"
                            placeholder="Enter a title for this card"
                            onChange={handleChangeColTitle}
                        />
                        <ButtonWIthIcon callback={addCard} color="#5aac44" text="Add" />
                      </TextAreaWrapper>
                  )}
                </React.Fragment>
            )}
          </BoardColumnWrapper>
      )}
    </Draggable>
  );
};
