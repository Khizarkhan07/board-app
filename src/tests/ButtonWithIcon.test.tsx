import * as React from "react";
import { shallow } from "enzyme";
import { ButtonWIthIcon } from "../components/ButtonWithIcon";
import {mount} from "enzyme";
import {Card} from "../components/card/Card";
import {Column} from "../components/column/Column";

test("Button has icon", () => {
  const Button = shallow(<ButtonWIthIcon icon={"fa fa-plus"} />);

  expect(Button.childAt(0).hasClass("fa fa-plus")).toBe(true);
  expect(Button).toMatchSnapshot();
});

test("Button has text", () => {
  const Button = shallow(<ButtonWIthIcon text={"inputText"} />);

  expect(Button.childAt(0).text()).toEqual("inputText");
});

test("Button has both icon and text", () => {
  const Button = shallow(
    <ButtonWIthIcon text={"inputText"} icon={"fa fa-plus"} />
  );

  expect(Button.childAt(1).text()).toEqual("inputText");
  expect(Button.childAt(0).hasClass("fa fa-plus")).toBe(true);
});

test("Button callback test", () => {
  let test = false;
  const Button = shallow(
    <ButtonWIthIcon
      callback={(e) => {
        test = true;
      }}
    />
  );
  Button.simulate("click");
  expect(test).toBe(true);
});
