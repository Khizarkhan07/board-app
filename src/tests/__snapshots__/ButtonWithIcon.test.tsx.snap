// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has icon 1`] = `
<button
  className="btn btn-light btn-outline mb-1"
  style={
    Object {
      "background": undefined,
      "width": "100%",
    }
  }
>
  <i
    className="fa fa-plus mr-1"
    style={
      Object {
        "color": "#fffff",
      }
    }
  />
</button>
`;
